onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx_common_int8.h

```diff
@@ -0,0 +1,758 @@
+
+  const std::byte* ablob = QuantARowPtr;
+  const std::byte* b_blk_data_ptr = QuantBDataColPtr;
+  const float* blk_scale_ptr = QuantBScaleColPtr;
+
+  [[maybe_unused]] size_t QuantBZeroPointIdx = 0;  // track half byte increments with this index instead of a pointer
+  // only used if HasZeroPoint == true
+
+  for (size_t k = 0; k < CountK; k += BlkLen) {
+    size_t ck = std::min(CountK - k, BlkLen);
+
+    const float a_scale = Q8BlkScale(ablob);
+    ablob += sizeof(float);
+
+    float
+      scale_v0 = (*(blk_scale_ptr + StrideQuantBScale * 0)) * a_scale,
+      scale_v1 = (*(blk_scale_ptr + StrideQuantBScale * 1)) * a_scale,
+      scale_v2 = (*(blk_scale_ptr + StrideQuantBScale * 2)) * a_scale,
+      scale_v3 = (*(blk_scale_ptr + StrideQuantBScale * 3)) * a_scale;
+
+    const std::byte* bptr0 = (b_blk_data_ptr + StrideQuantBData * 0);
+    const std::byte* bptr1 = (b_blk_data_ptr + StrideQuantBData * 1);
+    const std::byte* bptr2 = (b_blk_data_ptr + StrideQuantBData * 2);
+    const std::byte* bptr3 = (b_blk_data_ptr + StrideQuantBData * 3);
+
+    uint8_t zp0, zp1, zp2, zp3;
+    if constexpr (HasZeroPoint) {
+      // TODO: this block causes near 30% of the computation.
+      bool is_lower = (QuantBZeroPointIdx & 1) == 0;
+      std::byte zp_packed = QuantBZeroPointColPtr[0 * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+      zp0 = std::to_integer<int8_t>(is_lower ? (zp_packed & std::byte{ 0x0F }) : (zp_packed >> 4));
+      zp_packed = QuantBZeroPointColPtr[1 * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+      zp1 = std::to_integer<int8_t>(is_lower ? (zp_packed & std::byte{ 0x0F }) : (zp_packed >> 4));
+      zp_packed = QuantBZeroPointColPtr[2 * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+      zp2 = std::to_integer<int8_t>(is_lower ? (zp_packed & std::byte{ 0x0F }) : (zp_packed >> 4));
+      zp_packed = QuantBZeroPointColPtr[3 * StrideQuantBZeroPoint + QuantBZeroPointIdx / 2];
+      zp3 = std::to_integer<int8_t>(is_lower ? (zp_packed & std::byte{ 0x0F }) : (zp_packed >> 4));
+    } else {
+      zp0 = 8; zp1 = 8; zp2 = 8; zp3 = 8;
+    }
+
+    for (size_t kk = 0; kk < ck; kk += SubBlkLen64) {
+      // Load A row vector
+      const __m256i av0_32_epi8 = _mm256_loadu_si256((const __m256i*)ablob); ablob += 32;
+      const __m256i av1_32_epi8 = _mm256_loadu_si256((const __m256i*)ablob); ablob += 32;
+
+      // Load B column vectors (quantized to int4 blobs)
+      // dst: | v0  v32 | v1  v33 | ... | v30 v62 | v31 v63 |
+      __m256i bv = _mm256_loadu_si256((__m256i const*)bptr0); bptr0 += SubBlkStep;
+      __m256i bv0_32_epi8 = _mm256_and_si256(bv, low_mask);
+      __m256i bv1_32_epi8 = _mm256_and_si256(_mm256_srli_epi16(bv, 4), low_mask);
+      __m256i zp_epi8 = _mm256_set1_epi8(zp0);
+      bv0_32_epi8 = _mm256_sub_epi8(bv0_32_epi8, zp_epi8);
+      bv1_32_epi8 = _mm256_sub_epi8(bv1_32_epi8, zp_epi8);
+      __m256 sum_ps = dot_quad(bv0_32_epi8, bv1_32_epi8, av0_32_epi8, av1_32_epi8);
```

Comment:
is this able to be inlined if DotQuadFunctionType is a function pointer?

Reply:
added comment for further investigation